### PR TITLE
[DNNL] Softmax mkldnn op inplace support

### PR DESCRIPTION
**Pull request by jczaja**
This is initial support for in-place support of DNNL integration as disscussed in #23074. This PR is just introducing in-place support for softmax mkl-dnn op. In-place support to other ops: activation, batch-norm, sum, elemetnwise add will be added in next PRs as well as C-API pass.

### CODE REVIEW

**Comment by GaoWei8** [comment]
@jczaja  , The failed log is:
```
[14:10:26]	[Step 1/1] ****************
[14:10:26]	[Step 1/1] 0. The error message you wrote in PADDLE_ENFORCE{_**} or PADDLE_THROW does not meet our error message writing specification. Possible errors include 1. the error message is empty / 2. the error message is too short / 3. the error type is not specified. Please read the specification [ https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification ], then refine the error message. If it is a mismatch, please specify chenwhql (Recommend), luotao1 or lanxianghit review and approve.
[14:10:26]	[Step 1/1] The PADDLE_ENFORCE{_**} or PADDLE_THROW entries that do not meet the specification are as follows:
[14:10:26]	[Step 1/1] PADDLE_ENFORCE_EQ(&out_tensor, x);
[14:10:26]	[Step 1/1] 
[14:10:26]	[Step 1/1] There are 1 approved errors.
[14:10:26]	[Step 1/1] ****************
[14:10:35]	[Step 1/1] Process exited with code 1
[14:10:35]	[Step 1/1] Process exited with code 1 (Step: Build and test (Command Line))
[14:10:35]	[Step 1/1] Step Build and test (Command Line) failed
```

Please rewrite PADDLE_ENFORCE as [Paddle Error Message Writing Specification](https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification-(English-Verison)).

**Comment by jczaja** [comment]
@GaoWei8 Thanks for providing me a log snippet

**Comment by grygielski** [comment]
Looks great! But I've got few questions:

- Does this work automatically with Paddle's inplace mechanism? I don't see any graph/op modifications here so I assume it works out of the box for any model, am I right?
- How does it work in python API? Can we somehow force paddle to use inplace operator inside python program? If it's possible maybe it would be good to add python UTs too.
- Didn't you consider making `TestMain` function more general? For example with adding additional parameter something like `TestMain(const platform::Place &place, std::string operator, const framework::DDim &dims)` in this way we could avoid duplicating some code in future operators with inplace support using e.g. `auto op_ref = framework::OpRegistry::CreateOp(
      operator, input_map, output_ref_map, attrs);`. Ofc different operators can have different parameters so it should be handler inside by adding these things to specific maps. It's just an idea.

**Comment by jczaja** [comment]
@grygielski
1. This PR introduce in-place support to softmax mkl-dnn. To have it working with C-API there is enabling pass needed (WIP). It should work with python API via parallel executor which has in-place enabled by default (See #23074 for more info).
2. Thanks for suggesstions on making MainTest more general, I will use them when We are to add in-place for other ops e.g. sum ,elemetwise_add and activations.

**Comment by jczaja** [comment]
@luotao1 ,@GaoWei8 could you please reurn PR_Windows_CI ?

**Comment by GaoWei8** [comment]
> @luotao1 ,@GaoWei8 could you please reurn PR_Windows_CI ?

Done.

**Comment by jczaja** [comment]
@GaoWei8 Could you please double check if rerun of PR_Windows_CI(Paddle Windows) was made? I can see that build number did not change and from logs , execution date is still 2 days old
Finished **2020-03-25 16:50:16** with status FAILURE 'Exit code 8 (Step: test (Command Line)) (new)'

**Review by grygielski — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
@jczaja I rerun it and it is pending now.

**Comment by jczaja** [comment]
@luotao1 , @GaoWei8 Looks like  again I got some random failures . Could you please rerun PR_Windows_CI again ?

**Comment by zhwesky2010** [comment]
> @luotao1 , @GaoWei8 Looks like again I got some random failures . Could you please rerun PR_Windows_CI again ?

PR_Windows_CI has passed.

File: paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc [hunk 1]
```diff
@@ -93,8 +93,11 @@ class SoftmaxMKLDNNKernel : public paddle::framework::OpKernel<T> {
                                     ctx.GetPlace(), ctx.OutputName("Out"));
 
     auto softmax_src_memory_p = handler.AcquireSrcMemory(input);
-    auto softmax_dst_memory_p = handler.AcquireDstMemory(output);
     auto softmax_p = handler.AcquireForwardPrimitive();
+    // For Inplace src and and dst are the same memory object
+    auto softmax_dst_memory_p = input->Holder() == output->Holder()
```
Comment (luotao1) [inline paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc:98]: How about using the `IsSharedBufferWith` method in #23175 in the next PR?

**Review by luotao1 — APPROVED** [review]
